File: Data Structures & Algorithms/validate-parentheses/submission-2.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        let symbolDict = new Map([
            ['(', ')'],
            ['[', ']'],
            ['{', '}'],
        ])

        let stack = []

        if(s.length % 2 == 1) {
            return false
        }

        for(let char of s) {
            if(char === '(' ||
            char === '{' ||
            char === '[') {
                stack.push(char)
            }
            else if(char !== symbolDict.get(stack.pop())) {
                return false
            }
        }
        return stack.length == 0


    
        
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-2.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        out = []
        stack = []

        def backtrack (open_count, closed_count):
            if open_count == closed_count == n:
                out.append("".join(stack))
                return

            if open_count < n:
                stack.append("(")
                backtrack(open_count + 1, closed_count)
                stack.pop()
            
            if closed_count < open_count:
                stack.append(")")
                backtrack(open_count, closed_count + 1)
                stack.pop()

        backtrack(0,0)
        return out